src/NuGet.Core/NuGet.Protocol/Plugins/PluginManager.cs

```diff
@@ -150,60 +159,109 @@ public void Dispose()
+        /// </summary>
+        /// <param name="result">plugin discovery result</param>
+        /// <param name="requestedOperationClaim">The requested operation claim</param>
+        /// <param name="requestKey">plugin request key</param>
+        /// <param name="packageSourceRepository">package source repository</param>
+        /// <param name="serviceIndex">service index</param>
+        /// <param name="cancellationToken">cancellation token</param>
+        /// <returns>A plugin creation result</returns>
+        private async Task<PluginCreationResult> CreatePluginAsync(
+            PluginDiscoveryResult result,
+            OperationClaim requestedOperationClaim,
+            PluginRequestKey requestKey,
+            string packageSourceRepository,
+            JObject serviceIndex,
+            CancellationToken cancellationToken)
         {
             PluginCreationResult pluginCreationResult = null;
+            var cacheEntry = new PluginCacheEntry(_pluginsCacheDirectory.Value, result.PluginFile.Path, requestKey.PackageSourceRepository);
 
-            if (result.PluginFile.State.Value == PluginFileState.Valid)
-            {
-                var plugin = await _pluginFactory.GetOrCreateAsync(
-                    result.PluginFile.Path,
-                    PluginConstants.PluginArguments,
-                    new RequestHandlers(),
-                    _connectionOptions,
-                    cancellationToken);
-
-                var utilities = await PerformOneTimePluginInitializationAsync(plugin, cancellationToken);
-
-                var lazyOperationClaims = _pluginOperationClaims.GetOrAdd(
-                        requestKey,
-                        key => new Lazy<Task<IReadOnlyList<OperationClaim>>>(() =>
-                        GetPluginOperationClaimsAsync(
-                            plugin,
-                            packageSourceRepository,
-                            serviceIndex,
-                            cancellationToken)));
-
-                await lazyOperationClaims.Value;
-
-                pluginCreationResult = new PluginCreationResult(
-                    plugin,
-                    utilities.Value,
-                    lazyOperationClaims.Value.Result);
-            }
-            else
-            {
-                pluginCreationResult = new PluginCreationResult(result.Message);
-            }
+            return await ConcurrencyUtilities.ExecuteWithFileLockedAsync(
+                cacheEntry.CacheFileName,
+                action: async lockedToken =>
+                {
+                    cacheEntry.LoadFromFile();
+                    if (cacheEntry.OperationClaims == null || cacheEntry.OperationClaims.Contains(requestedOperationClaim))
+                    {
+                        if (result.PluginFile.State.Value == PluginFileState.Valid)
```

Comment:
If the plugin file state itself isn't valid, then do you need to even go through the cache entry?

Reply:
The plugin file state might involve an authenticode verification. 
I changed that in this PR. https://github.com/NuGet/NuGet.Client/pull/2266
Checking a small file cache is cheaper than doing authenticode verification :)

Reply:
`Checking a small file cache is cheaper than doing authenticode verification`
Not sure, authenticode verification will happen only once but cache file read will happen every single time. and what do we know disk i/o isn't cheap :) 

May be you can check if `result.PluginFile.State` value is already created then it's not valid then just return default PluginResult?

Reply:
The cache file read will happen 1x per operation claim. 
It's a minimal disk hit, compared to authenticode which will involve remoting. 
It's a ms vs seconds difference. 

I could check if the authenticode check is evaluated, but that would make the rest of initialization more complicated and potentially more expensive.